.github/workflows/compile-requirements.yml

```diff
@@ -0,0 +1,137 @@
+        description: An optional dependency installation specification for use with a pyproject.toml file.
+        type: string
+        required: false
+        default: ".[dev]"
+      upstream-packages:
+        description: A whitespace separated list of Howso package dependencies (e.g., "amalgam-lang-py howso-engine-py").
+        type: string
+        required: false
+
+defaults:
+  run:
+    shell: bash
+
+jobs:
+  check-requirements:
+    runs-on: ubuntu-latest
+    outputs:
+      dev_req_file: ${{ steps.detect-changes.outputs.dev-req-file }}
+    steps:
+      - uses: actions/checkout@v4
+        with:
+          ref: ${{ github.head_ref }}
+
+      - name: Check for requirements changes
+        id: detect-changes
+        run: |
+          # Determine which type of requirements/setup file exists in this repo
+          if [ -e "pyproject.toml" ]; then
+            echo "Found pyproject.toml"
+            req_file="pyproject.toml"
+          elif [ -e "requirements.in" ]; then
+            echo "Found requirements.in"
+            req_file="requirements.in"
+            if [ -e "requirements-dev.in" ]; then
+              dev_req_file="requirements-dev.in"
+            fi
+          fi
+
+          # Check for changes in myfile.txt on the current branch compared to main branch
+          git fetch
+          git status
+          changes=$(git diff "origin/main" --name-only -- "$req_file")
+
+          if [ -n "$changes" ]; then
+              echo "Changes found in $req_file on branch ${{ github.head_ref }}."
+          else
+              echo "No changes to $req_file detected on ${{ github.head_ref }}."
+          fi
+
+          echo "dev-req-file: $dev_req_file"
+
+          echo "dev-req-file=$(echo $dev_req_file)" >> $GITHUB_OUTPUT
+
+  compile-requirements:
+    needs: ['check-requirements']
+    runs-on: ubuntu-latest
+    permissions:
+      contents: write
+    outputs:
+      changes-pushed: ${{ steps.commit-changes.outputs.changes-pushed }}
+    strategy:
+      max-parallel: 1
+      matrix:
+        type: ['3.8', '3.9', '3.10', '3.11', 'licenses']
+    steps:
+
+      - uses: actions/checkout@v4
+        with:
+          token: ${{ secrets.HOWSOAI_WORKFLOW_AUTOMATION_TOKEN }}
+          ref: ${{ github.head_ref }}
+
+      - name: Set up Python
+        if: matrix.type != 'licenses'
+        uses: actions/setup-python@v4
+        with:
+          python-version: ${{ matrix.type }}
+
+      - name: Set up Python (3.11 only)
+        if: matrix.type == 'licenses'
+        uses: actions/setup-python@v4
+        with:
+          python-version: "3.11"
+
+      - name: Compile requirements
+        if: matrix.type != 'licenses'
+        id: compile-requirements
+        run: |
+          # If the previous commit was to compile requirements for the current version, skip
+          if [[ "$(git log --pretty=%B -n 1)" == "Automated requirements/license generation [${{ matrix.type }}]" ]]; then
+            exit 0
+          fi
```

Comment:
This check in essence is good but I don't think it actually works as expected. It does skip if the exact previous commit was the same version but that only happens once per run through. So, the 'lagging' commits (n-2,n-3....) will still be run even though they have already been run.

There won't be any updates for those previous runs but it will still make it through this check which just slows the run down.